### PR TITLE
feat: support suffix decoding for llm.

### PR DESCRIPTION
**Pull request by Clement-Wang26**
details:
- Add SuffixWorkerImpl for suffix-based speculative decoding.
- Add suffix tree and suffix decoding cache utilities.
- Add suffix-related flags and options.
- Suffix decoding uses greedy sampling for validation.
- worker.cpp factory dispatches 'suffix' algorithm to SuffixWorkerImpl.

rewards:
- TPOT improved by ~38.6% on the agentic_code_dataset_22 dataset compared to baseline without suffix decoding.
https://huggingface.co/datasets/novita/agentic_code_dataset_22/tree/main

### CODE REVIEW

File: xllm/core/runtime/suffix_worker_impl.cpp [hunk 1]
```diff
@@ -0,0 +1,438 @@
+/* Copyright 2025 The xLLM Authors. All Rights Reserved.
```
Comment (XuZhang99) [inline xllm/core/runtime/suffix_worker_impl.cpp]: 2026

File: xllm/core/runtime/suffix_worker_impl.cpp [hunk 2]
```diff
@@ -0,0 +1,438 @@
...
+    out += ",...";
+  }
+  out += "]";
+  return out;
+}
+
+}  // namespace
+
+namespace {
+runtime::Options SuffixTargetOptions(const runtime::Options& options) {
```
Comment (XuZhang99) [inline xllm/core/runtime/suffix_worker_impl.cpp:59]: rename this func name and move it to the above namespace.

File: xllm/core/util/suffix_decoding_cache.h [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+  int match_len = 0;
+
+  static SuffixDecodingDraft from_native(const Draft& draft);
+};
+
+class SuffixDecodingCache {
+ public:
+  SuffixDecodingCache(int max_tree_depth = 64, int max_cached_requests = -1);
+
+  int max_tree_depth() const { return max_tree_depth_; }
```
Comment (XuZhang99) [inline xllm/core/util/suffix_decoding_cache.h]: `int` ==> `int32_t` or `int64_t`

File: xllm/core/util/suffix_tree.cpp [hunk 4]
```diff
@@ -0,0 +1,949 @@
...
+    assert(idx < node->length);
+    if (ref_data[idx] != token) {
+      return {nullptr, -1};
+    }
+    idx++;
+  }
+  return {node, idx};
+}
+
+Draft SuffixTree::_speculate_path(Node* node,
```
Comment (XuZhang99) [inline xllm/core/util/suffix_tree.cpp]: remove `_` at the begin of func name.

Comment (Clement-Wang26) [reply xllm/core/util/suffix_tree.cpp]: done.

File: xllm/core/util/suffix_decoding_cache.h (hunk 3 above)
Comment (Clement-Wang26) [reply xllm/core/util/suffix_decoding_cache.h]: done. int -> int32_t now.

File: xllm/core/runtime/suffix_worker_impl.cpp (hunk 1 above)
Comment (Clement-Wang26) [reply xllm/core/runtime/suffix_worker_impl.cpp]: done.

File: xllm/core/common/global_flags.h [hunk 5]
```diff
@@ -66,6 +66,20 @@ DECLARE_int32(num_speculative_tokens);
 DECLARE_string(speculative_algorithm);
 DECLARE_bool(enable_opt_validate_probs);
 
+DECLARE_string(speculative_algorithm);
+
```
[the remark this answers is not in the document]
Comment (Clement-Wang26) [reply xllm/core/common/global_flags.h:70]: done.

File: xllm/core/distributed_runtime/speculative_engine.cpp [hunk 6]
```diff
@@ -33,6 +34,8 @@ SpeculativeEngine::SpeculativeEngine(const runtime::Options& options)
   CHECK_GT(options.num_speculative_tokens(), 0)
       << "speculative tokens should not be zero";
 
+  use_draft_engine_ = options_.speculative_algorithm() != "Suffix";
```
Comment (yq33victor) [inline xllm/core/distributed_runtime/speculative_engine.cpp]: Use a subclass to inherit SpeculativeEngine

File: xllm/core/runtime/forward_params.h [hunk 7]
```diff
@@ -156,6 +158,8 @@ struct ForwardInput {
 
   // step-level decode metadata
   std::optional<StepDecodeMeta> step_decode;
+  // If true, skip sampler forward and only keep logits.
+  bool skip_sampling_for_logits_only = false;
```
Comment (yq33victor) [inline xllm/core/runtime/forward_params.h:162]: when we need  this

File: xllm/core/runtime/llm_worker_impl.cpp [hunk 8]
```diff
@@ -148,28 +148,29 @@ std::optional<ForwardOutput> LLMWorkerImpl::step(const ForwardInput& input) {
   }
 
   // driver prepare model output
-  SampleOutput sample_output;
   if (sampling_params.selected_token_idxes.defined()) {
-    sample_output = sampler_->forward(logits, sampling_params);
```
Comment (yq33victor) [inline xllm/core/runtime/llm_worker_impl.cpp:153]: Becareful, `logits` will be modified in `sampler_->forward`, so we need to check the modification which move `sampler_->forward` after `output.logits = logits;`.

Comment (Clement-Wang26) [reply xllm/core/runtime/llm_worker_impl.cpp:153]: Actually, suffix decoding doesn't need to apply any sampling parameters to modify logits — it's pointless since we only care about the token with the highest probability.

File: xllm/core/runtime/forward_params.h (hunk 7 above)
Comment (Clement-Wang26) [reply xllm/core/runtime/forward_params.h:162]: enable suffix decoding.

File: xllm/core/distributed_runtime/speculative_engine.cpp (hunk 6 above)
Comment (Clement-Wang26) [reply xllm/core/distributed_runtime/speculative_engine.cpp]: done.

**Review by XuZhang99 — APPROVED** [review]
